mllib/src/main/scala/org/apache/spark/ml/classification/LogisticRegression.scala

```diff
@@ -370,49 +420,102 @@ class LogisticRegression @Since("1.2.0") (
+                val featureIndex = if ($(fitIntercept)) {
+                  index % numFeaturesPlusIntercept
+                } else {
+                  index % numFeatures
+                }
                 // If `standardization` is false, we still standardize the data
                 // to improve the rate of convergence; as a result, we have to
                 // perform this reverse standardization by penalizing each component
                 // differently to get effectively the same objective function when
                 // the training dataset is not standardized.
-                if (featuresStd(index) != 0.0) regParamL1 / featuresStd(index) else 0.0
+                if (featuresStd(featureIndex) != 0.0) {
+                  regParamL1 / featuresStd(featureIndex)
+                } else {
+                  0.0
+                }
               }
             }
           }
           new BreezeOWLQN[Int, BDV[Double]]($(maxIter), 10, regParamL1Fun, $(tol))
         }
 
         val initialCoefficientsWithIntercept =
-          Vectors.zeros(if ($(fitIntercept)) numFeatures + 1 else numFeatures)
-
-        if (optInitialModel.isDefined && optInitialModel.get.coefficients.size != numFeatures) {
-          val vecSize = optInitialModel.get.coefficients.size
-          logWarning(
-            s"Initial coefficients will be ignored!! As its size $vecSize did not match the " +
-            s"expected size $numFeatures")
+          Vectors.zeros(numCoefficientSets * numFeaturesPlusIntercept)
+
+        val initialModelIsValid = optInitialModel.exists { model =>
+          val providedCoefs = model.coefficientMatrix
+          val modelValid = (providedCoefs.numRows == numCoefficientSets) &&
+            (providedCoefs.numCols == numFeatures) &&
+            (model.interceptVector.size == numCoefficientSets)
+          if (!modelValid) {
+            logWarning(s"Initial coefficients will be ignored! Its dimensions " +
+              s"(${providedCoefs.numRows}, ${providedCoefs.numCols}) did not match the expected " +
+              s"size ($numCoefficientSets, $numFeatures)")
+          }
+          modelValid
         }
 
-        if (optInitialModel.isDefined && optInitialModel.get.coefficients.size == numFeatures) {
-          val initialCoefficientsWithInterceptArray = initialCoefficientsWithIntercept.toArray
-          optInitialModel.get.coefficients.foreachActive { case (index, value) =>
-            initialCoefficientsWithInterceptArray(index) = value
+        if (initialModelIsValid) {
+          val initialCoefArray = initialCoefficientsWithIntercept.toArray
+          val providedCoef = optInitialModel.get.coefficientMatrix
+          providedCoef.foreachActive { (row, col, value) =>
+            val flatIndex = row * numFeaturesPlusIntercept + col
+            // We need to scale the coefficients since they will be trained in the scaled space
+            initialCoefArray(flatIndex) = value * featuresStd(col)
```

Comment:
Cool. I think the original code doesn't do `value * featuresStd(col)` which is a bug. Thank you for finding it. Can you change `initialCoefArray` into something withIntercept? Also, please check if the parent initalModel has fitIntercept or not.